sql/expression/function/aggregation/window/lag.go

```diff
@@ -0,0 +1,264 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package window
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"gopkg.in/src-d/go-errors.v1"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+)
+
+var ErrInvalidLagOffset = errors.NewKind("'LAG' offset must be a non-negative integer; found: %v")
+var ErrInvalidLagDefault = errors.NewKind("'LAG' default must be a literal; found: %v")
+
+type Lag struct {
+	window *sql.Window
+	expression.NaryExpression
+	//def sql.Expression
+	offset int
+	pos    int
+}
+
+var _ sql.FunctionExpression = (*Lag)(nil)
+var _ sql.WindowAggregation = (*Lag)(nil)
+
+// getLagOffset extracts a non-negative integer from an expression.Literal, or errors
+func getLagOffset(e sql.Expression) (int, error) {
+	lit, ok := e.(*expression.Literal)
+	if !ok {
+		return 0, ErrInvalidLagOffset.New(e)
+	}
+	val := lit.Value()
+	var offset int
+	switch e := val.(type) {
+	case int:
+		offset = e
+	case int8:
+		offset = int(e)
+	case int16:
+		offset = int(e)
+	case int32:
+		offset = int(e)
+	case int64:
+		offset = int(e)
+	default:
+		return 0, ErrInvalidLagOffset.New(e)
+	}
+
+	if offset < 0 {
+		return 0, ErrInvalidLagOffset.New(e)
+	}
+
+	return offset, nil
+}
+
+// NewLag accepts variadic arguments to create a new Lag node:
+// If 1 expression, use default values for [default] and [offset]
+// If 2 expressions, use default value for [default]
+// 3 input expression match to [child], [offset], and [default] arguments
+// The offset is constrained to a non-negative integer expression.Literal.
+func NewLag(e ...sql.Expression) (*Lag, error) {
+	switch len(e) {
+	case 1:
+		return &Lag{NaryExpression: expression.NaryExpression{ChildExpressions: e[:1]}, offset: 1}, nil
+	case 2:
+		offset, err := getLagOffset(e[1])
+		if err != nil {
+			return nil, err
+		}
+		return &Lag{NaryExpression: expression.NaryExpression{ChildExpressions: e[:1]}, offset: offset}, nil
+	case 3:
+		offset, err := getLagOffset(e[1])
+		if err != nil {
+			return nil, err
+		}
+		return &Lag{NaryExpression: expression.NaryExpression{ChildExpressions: []sql.Expression{e[0], e[2]}}, offset: offset}, nil
+	}
+	return nil, sql.ErrInvalidArgumentNumber.New("LAG", "1, 2, or 3", len(e))
+}
+
+// Description implements sql.FunctionExpression
+func (l *Lag) Description() string {
+	return "returns value of argument from first row of window frame."
+}
+
+// Window implements sql.WindowExpression
+func (l *Lag) Window() *sql.Window {
+	return l.window
+}
+
+// IsNullable implements sql.Expression
+func (l *Lag) Resolved() bool {
+	return windowResolved(l.window)
```

Comment:
The default also needs to be resolved, no?

Reply:
yep